src/testing/adebug.h

```diff
@@ -11,10 +11,26 @@
     #define FUNC_IDENT __func__
 #endif
 
+// CAUTION: qDebug() doesnt print for non-DEBUG builds
 #define DEB qDebug() << FUNC_IDENT << "\n\t"
 #define DEB_SRC DEB
 #define DEB_RAW qDebug() << '\t'
 
+
```

Comment:
I prefer to stick with DEB for Debugging and LOG for console output to the end user for now, keep it simple. LOG can be used for INFO, WARN and CRIT type messages, so no need to add another 3 additional 'levels' of console output. We can look into colour coding different LOG levels, but I wouldn't make it a priority right now.

I think the NOT_IMPLEMENTED macro is a good and useful addition for WIP sections!

Reply:
The idea is that Info Warn Crit are actually part of the application. Its more or less what the user would copy paste as "error" messages for us to figure out. DEB and so qDebug() is literally just that developement debug so while we can push it back the distinction should be established and appropriately used.
For example any necessary info like actions pressed, switching logic due to some condition etc that when the actual program is running is necessary should _not_ be debug. However internal variable values (pointers etc) can be debug

Reply:
Hmm, since we are building a GUI application here, the vast majority of users is going to run the app from a start menu / app launcher or similar, meaning they never get to see any console output anyway, regardless of whether it's qDebug, qWarning or anything else.

If we are interested in bug / crash reports, those would have to be written to a text file anyway before being sent to us, so that any colourful coding or differentiating between how messages are displayed would be moot.

LOG can provide all we need in this case, which is logging to console (still need to implement the logging to file). One of the reasons why I'm opposed to differentiating it further is that it opens up a whole range of discussions about is this more suitable for WARN or for CRIT or for INFO. Or even if not discussions, it's still another mental step when producing console out, plus it makes the code more convoluted to read and write.

I'm a big fan of the [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and try to keep it as a guiding principle for this project, so let's keep it simple stupid ;)

Reply:
Dealing with the color is engineering not a problem.
![image](https://user-images.githubusercontent.com/69913906/112059942-d1520e80-8b5c-11eb-82d4-952d35ae1e8c.png)

At the same time there are levels of output the application outputs. Logging everything as LOG is _too_ simple. Yes keep it simple but not simpler should be the moto.

Check any logging api you can find they all categorise output. Many in much more extensive ways than info warn crit. As they should bcause thats how a program runs. 

The discussion of what is what (info or warn) is unavoidable and should follow some convention based on the devs' criteria.

Reply:
Most logging API's have different log levels to differentiate between the urgency of the messages that are being displayed, with the rationale being that higher importance messages should catch the attention of the user/sysadmin etc. This means that in a text file, or a status console, messages are printed in different colours, or with different headers, to make them differentiable. This makes a lot of sense in those scenarios, obviously.

However, in a GUI application we need to handle this differently. Anything that should be a "Warning" that should let's say catch the attention of the user (say, an invalid input) has to be displayed in the GUI, as a `QMessageBox` for example, since the end user doesn't see the console output - he's not even running a terminal. What good does a console warning about an invalid input do if the user never gets to see it?

That being said, for hunting down bugs and pinpointing where something went wrong, logging milestones is definitely helpful. I will also agree that having normal 'milestones' like a new flight being logged successfully being differentiated from 'non-normal' returns in functions is a sensible thing, so having different LOG levels is something I am willing to implement.

However, let's take this step by step and not get distracted by it now. I will look into setting up logging and we can discuss the implementation once you are finished with the backup widget. Let's stick with DEB and LOG for now and I will have something ready by the time the backup widget is finished!